### PR TITLE
Factor out pooling and lifetime management logic of DHO to a base class

### PR DESCRIPTION
**Pull request by ekrctb**
- [x] Depends on #12486

To implement correct and efficient replay rewinding behavior for a gameplay element, some information has to be stored in a non-Drawable type (I call this just "entry") and the entries are recorded to a list, and a pooled drawable is realized when the current time is in the lifetime of the entry.
This is currently implemented in `HitObjectLifetimeEntry` (entry) and `DrawableHitObject` (drawable).

However, there are cases when something not really a "hit object" wants to use the lifetime management logic above. To list,

- `DrawableBarLine` in mania/taiko: are `DrawableHitObject`, but with the awkward special code in `TaikoPlayfield`.
- `HitExplosion` in mania/catch: doesn't rewind correctly (unnoticed because of the short lifetime).
- `DrawableFlyingHit` in taiko: also requires special handling, and the special `IgnoreHit` dummy hit object.
- `CaughtObject` in catch: doesn't rewind correctly. I tried to implement lifetime management logic in #11255, but it ended up with a big chunk of mess.

In this PR, I factor out the lifetime management logic to the new `DrawableObject` class and making it the base class of DHO.
Non-HitObject things like in the list above will be implemented using the new class, correct and efficient replay rewinding behavior in a minimal implementation.
What is not implemented yet in this PR is a `HitObjectContainer` generalization, and the new `DrawableObject` is not easily usable yet. I made this PR to get comments on the design.

I removed the `SkinReloadableDrawable` inheritance from DHO, a minor breaking change. I think the value of `SkinReloadableDrawable` is minimal as the required code to support the skinning feature is small. `SkinnableDrawableObject` or something could be added later if it would be worth it.

### CODE REVIEW

**Comment by ekrctb** [comment]
I rebased to the merged version of the PR (changes are made after I branched).

**Comment by bdach** [comment]
A normal merge would have been just as fine if not better. It's not standard practice around here to rebase over dependencies even if they're heavily changed. The commits will disappear from the GitHub PR view anyways.

File: osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs [hunk 1]
```diff
@@ -227,22 +207,15 @@ public void Apply([NotNull] HitObject hitObject)
...
             // We override this with DHO's InitialLifetimeOffset for a non-pooled DHO.
-            if (newEntry is SyntheticHitObjectEntry)
-                lifetimeEntry.LifetimeStart = HitObject.StartTime - InitialLifetimeOffset;
+            if (entry is SyntheticHitObjectEntry)
+                entry.LifetimeStart = HitObject.StartTime - InitialLifetimeOffset;
 
-            LifetimeStart = lifetimeEntry.LifetimeStart;
-            LifetimeEnd = lifetimeEntry.LifetimeEnd;
+            LifetimeStart = entry.LifetimeStart;
+            LifetimeEnd = entry.LifetimeEnd;
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:218]: Isn't this redundant now (already done in base)?

File: osu.Game/Rulesets/Objects/Pooling/DrawableObject.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+using osu.Framework.Graphics.Performance;
+using osu.Framework.Graphics.Pooling;
+
+namespace osu.Game.Rulesets.Objects.Pooling
+{
+    /// <summary>
+    /// A <see cref="PoolableDrawable"/> that is controlled by <see cref="Entry"/> to implement drawable pooling and replay rewinding.
+    /// </summary>
+    /// <typeparam name="TEntry">The <see cref="LifetimeEntry"/> type storing state and controlling this drawable.</typeparam>
+    public abstract class DrawableObject<TEntry> : PoolableDrawable where TEntry : LifetimeEntry
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/Pooling/DrawableObject.cs:16]: Needs a better name. `DrawableObject` is too general for what this is. `DrawableObjectWithLifetime`, maybe. Or `PoolableDrawableWithLifetime`...?

This is generic enough that you could probably even put it in framework at this rate.

File: osu.Game/Rulesets/Objects/Pooling/DrawableObject.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+            HasEntryApplied = true;
+        }
+
+        protected sealed override void FreeAfterUse()
+        {
+            base.FreeAfterUse();
+
+            if (IsInPool)
+                freeIfInUse();
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/Pooling/DrawableObject.cs:78]: This is weird. One guard is out here (`IsInPool`) and the other is inside (`HasEntryApplied`). They either need to be both outside or inside.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
Aside from the above:

* With regard to removing skinnability: I'm not sure the removal of `SkinReloadableDrawable` is so minor. The `allowFallback` stuff is now hard-coded to `true`. Are you sure that isn't breaking anything?
* Again, as previously, would like to see how the components you named would eventually look with this design.
* Another pain point with pooled drawables is having to manage transforms. We have several places where similar logic of "rewind transforms to negative infinity, clear transforms, reapply new ones" is done. The new `PoolableDrawableWithLifetime` may be a decent place to move that to.

**Comment by peppy** [comment]
This is going to require a pass to figure out if we actually need `allowFallback` in the first place. On a quick check it's never been used (yet), but may be part of the puzzle of getting legacy-legacy fallback working (as tracked in #9863).

File: osu.Game/Rulesets/Objects/Pooling/DrawableObject.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/Pooling/DrawableObject.cs:16]: I like `PoolableDrawableWithLifetime`. Hopefully this class can be used for other things like followpoints?

**Comment by ekrctb** [comment]
As peppy said, the features of `SkinReloadableDrawable` base class were largely unused, and it is why I remove it here.
We can add back the inheritance layer in the future if it is proved to be worth having it.

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**